jinja2 for fixes, checks and the opencontrol yaml

mpreisler opened the pull request:
As SSG grows bigger we feel the need for a more advanced and unified templating system. I am hoping jinja2 will become that. It's standard, has great docs and a lot of examples online.

This PR processes every check, fix and opencontrol yaml with jinja2. It provides 2 examples, one for OVAL checks and one for opencontrol yaml. So far we have no fix using the jinja2 capabilities but I am hoping that will change in time once this PR gets merged.

pep8speaks commented:
Hello @mpreisler! Thanks for submitting the PR.



 - In the file [`shared/utils/combine-remediations.py`](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py), following are the PEP8 issues :

> [Line 21:1](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L21): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 22:1](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L22): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 100:33](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L100): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 101:33](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L101): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 105:33](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L105): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 106:33](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L106): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 110:33](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L110): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 111:33](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L111): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 115:33](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L115): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 116:33](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L116): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 278:37](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L278): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 280:37](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L280): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 292:37](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L292): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 338:49](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L338): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 339:49](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L339): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 342:57](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L342): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 342:79](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L342): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 343:57](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L343): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 486:49](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L486): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 547:41](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L547): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 564:1](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/combine-remediations.py#L564): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1

 - In the file [`shared/utils/yaml-to-shorthand.py`](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/yaml-to-shorthand.py), following are the PEP8 issues :

> [Line 21:1](https://github.com/OpenSCAP/scap-security-guide/blob/4158d0b901005f2414f50c2b11fde9bed5cbbe3d/shared/utils/yaml-to-shorthand.py#L21): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file

mpreisler commented:
I was trying to find a good example to showcase the capabilities of this feature. Found some issues in SSG instead: #2786 and #2787

So far I can't find any good example for jinja2 in fixes. There definitely are cases where jinja2 would be great for OVALs but with fixes you can always use the language to do conditional execution - bash or ansible.

mpreisler commented:
(the jenkins failures are expected because jinja2 package is missing on our jenkins slaves, please don't let that keep you from looking at this PR :-))

mpreisler commented:
I will proceed to implement this for checks as well because that's where it will help more at the moment. I am sure we can use it later for fixes as well.

mpreisler commented:
> @mpreisler I assume that {% if product == 'rhel7' %} would be how the do product handling?

yes but I am hoping this will be rare. something like {{% if rpm_based %}} should be more common.

mpreisler commented:
@openscap-ci test this please

mpreisler commented:
TODO:
- provide an example of an OVAL check using jinja2
- provide an example of a bash fix using jinja2
- provide an example of opencontrol yaml using jinja2 - remote OVAL is a good candidate

Then I think it can be merged. Error reporting really sucks right now but that can be improved later.

mpreisler commented:
I have made sure jenkins workers have python-jinja2 installed. See https://github.com/OpenSCAP/jenkins/commit/f452139d3eadf058e44823c4a63b27c498bb053e

yuumasato commented on shared/guide/system/software/updating/security_patches_up_to_date.rule:
```diff
@@ -1,15 +1,23 @@
 title: 'Ensure Software Patches Installed'
 
 description: |-
+{{% if product in ["rhel6", "rhel7"] %}}
```
Syntax here looks good to me. I like that we can use this kind of construct.

yuumasato commented on shared/templates/template_OVAL_package_installed:
```diff
@@ -4,15 +4,25 @@
+  <linux:rpminfo_object id="obj_package_%PKGNAME%_installed" version="1">
+    <linux:name>%PKGNAME%</linux:name>
+  </linux:rpminfo_object>
+{{% elif pkg_system == "dpkg" %}}
```
Shouldn't this be `deb`?

mpreisler replied:
no. see https://en.wikipedia.org/wiki/Dpkg

deb is short for debian, dpkg is the packaging system

yuumasato replied:
Ok.

yuumasato commented on shared/templates/template_OVAL_package_installed:
```diff
@@ -4,15 +4,25 @@
+        <platform>multi_platform_all</platform>
       </affected>
-      <description>The DEB package %PKGNAME% should be installed.</description>
+      <description>The {{{ pkg_system|upper }}} package %PKGNAME% should be installed.</description>
```
But then here it be expanded to `DPKG package`, not `DEB package`.

yuumasato replied:
It may actually be `DPKG package`, but I think I never saw like this.

mpreisler replied:
is that a problem? if anything DPKG package is more correct because DEB is short for debian so it doesn't apply on Ubuntu.

yuumasato commented on shared/utils/combine-ovals.py:
```diff
@@ -278,18 +278,21 @@ def checks(product, oval_version, oval_dirs):
+                    if not check_is_applicable_for_product(
+                        xml_content,
+                        ssgcommon.required_yaml_key(product_yaml, "product")
+                    ):
```
Do you see jinja being used to determine applicability? Otherwise we can check for applicability before processing.

mpreisler replied:
It's not used for applicability right now but I am hoping it will replace the applicability checking in the future. We have to process first because the other order breaks parsing error functionality. It's doable but not without loading the file twice.

yuumasato replied:
Thanks for clarification.

mpreisler commented:
I have removed the weird relpath hack and implemented an absolute file path loader for jinja2. Submitted it as a PR, see https://github.com/pallets/jinja/pull/846

mpreisler commented:
I don't want to do that as part of this PR but in another PR I want to start using jinja2 macros to replace some of the templating and shared bash functionality. See http://davedash.com/tutorial/jinja-macros/

We could replace all package install and remove with jinja2 code that always puts in the correct minimal snippet into fixes.
